Adds Windows support and AppVeyor CI

0vercl0k opened the pull request:
Hey guys,

I've mentioned to you guys during BH that I'd have a look at how much effort it'd be to port miasm to Windows and so here it is.

This PR has all the changes required to have miasm working on Windows 32b/64b with Visual Studio (I've tested VS2017 but it'd be very easy to add different versions if you think it'd be useful in AppVeyor).

Here is the AppVeyor output for the below commit: https://ci.appveyor.com/project/0vercl0k/miasm-nc0e9. As you can see 460 tests out of the 471 of the Linux version are passing just fine.

I've disabled on purpose tests related to `dse_crackme.c` as the `dse_crackme.py` script is expecting an ELF to work properly. Adding support for a Windows PE or/and adding the ELF binary in the git tree would allow to have it tested on Windows too.

There is one last detail that I don't really know how to address though. Basically, if you guys know a way to have `setuptools` move extra files into the build directory (`.lib` files are necessary on Windows in order to have the jitter code being able to link against them). Right now, I manually copy them into the folder. The issue with that is that the content of the folder is moved into the installation directory prior to me moving the files. That's why in the `.appveyor.yml` file I run the install twice in a row; as the `.lib` files will be moved after the first install, the second actually places them into the installation directory. Not a huge deal, but clearly a hack and there must be a cleaner way of doing it :).

Worth mentioning: I've encountered at least two issues with `PyLong_FromLong` / `PyLong_FromUnsignedLong` and I'm unsure why those issues didn't show up on Linux. Maybe you guys have an idea?

One last thing, I've also disabled the `mn_div.py` test with the `llvm` jitter because it yields this on Windows:
```
LLVM ERROR: Program used external function '__umodti3' which could not be resolved!
```

And I'm not too sure who is responsible to expose this routine; again maybe you guys know :)

Happy to answer/discuss any questions you guys have!

Cheers

PS: Thanks to @commial  for helping out with questions and all 👍

commial commented:
Hi @0vercl0k,

First, a big thanks for the contrib!

>I've disabled on purpose tests related to dse_crackme.c as the dse_crackme.py script is expecting an ELF to work properly. Adding support for a Windows PE or/and adding the ELF binary in the git tree would allow to have it tested on Windows too.

We will include the `dse_crackme` compiled version, for test reproducibility and to avoid relying on the tester's compiler.
Also, we will likely extract the "run_with_linux_env" test from the test_all to put it only in the Travis CI testset. Indeed, it also relies on a few file / libs external to Miasm.

> One last thing, I've also disabled the mn_div.py test with the llvm jitter because it yields this on Windows:
> LLVM ERROR: Program used external function '__umodti3' which could not be resolved!

The LLVM jitter actually relies on a LLVM lib for a few symbols. On Linux, the corresponding library is loaded in the current process [here](https://github.com/cea-sec/miasm/blob/master/miasm2/jitter/llvmconvert.py#L169). This library comes with the LLVM's installation.

> PS: Thanks to @commial for helping out with questions and all +1

Please, do not hesitate :smiley:

serpilliere commented:
Ok, it seems travis doesn't accept our malware sample anymore: `sc_connect_back.bin`
I will do a PR to remove it. Once it's merged, your branch may be ok :smile:

serpilliere commented:
I will try #838 to fix the sample deletion.

0vercl0k commented:
Awesome :).

I still need to do some adjustments for the dse_crackme testcases for
Windows. I'll ping you guys once it's working and rebased properly :).

Cheers

Le jeu. 30 août 2018 à 06:15, serpilliere <notifications@github.com> a
écrit :

> I will try #838 <https://github.com/cea-sec/miasm/pull/838> to fix the
> sample deletion.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cea-sec/miasm/pull/835#issuecomment-417315018>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABaHReaNrOR0QqRZsEIfMPBGi4NFOLEHks5uV-VkgaJpZM4WRTL0>
> .
>

0vercl0k commented:
All right - I've cleaned up the last few issues. Basically `os.rename` behaves differently on Windows/Linux and I've added a work around to ignore concurrent `os.rename` (it was getting triggered when the three `dse_crackme` tests were being run).

With those modifications, it's green both on Travis and on AppVeyor: https://ci.appveyor.com/project/0vercl0k/miasm-nc0e9 & https://travis-ci.org/cea-sec/miasm/builds/422650577?utm_source=github_status&utm_medium=notification

```
...
Result: 462/462 pass
...
Creating archive: ..\miasm.x64.release.7z
Items to compress: 193
Files read from disk: 167
Archive size: 454136 bytes (444 KiB)
Everything is Ok
Collecting artifacts...
Found artifact 'miasm.x64.release.7z' matching 'miasm.*.7z' path
Uploading artifacts...
[1/1] miasm.x64.release.7z (454,136 bytes)...100%
Build success
```
```
...
Result: 470/470 pass
...
The command "python -c 'import z3; x = z3.BitVec(chr(0x41), 32)' && cd test && python test_all.py && git ls-files -o --exclude-standard" exited with 0.
```
Cheers

serpilliere commented:
Hi!

Ok for the work around: As far as I can see, it may fail in some cases, (if the first creator has created the file, but has not filled it with the code yet). But It's fine for me now :smiley:

For the squash, it's ok for me!
I am currently reviewing the whole patch.

0vercl0k commented:
Cool @serpilliere  - let me know what you think!

I've observed some failure on the DSE crackme that are pretty rare though (it looks like some sort of ~race) but I haven't been able to pinpoint it so far; for example from yesterday https://ci.appveyor.com/project/0vercl0k/miasm-nc0e9/build/1.8 on the x86 build:

```
ERROR dse_crackme.py ..\samples\dse_crackme --strategy branch-cov
[WARNING]: Create dummy entry for 'xxx.dll'
[INFO]: xxx___libc_start_main(main=0xa43, argc=0x2, ubp_av=0x13ff9d, init=0xac0, fini=0xb30, rtld_fini=0x0, stack_end=0x13ff88) ret addr: 0x73a
[INFO]: xxx_fopen(path=0x13ffc5, mode=0xd60) ret addr: 0x90a
[INFO]: xxx_fread(ptr=0x13fe30, size=0x1, nmemb=0x100, stream=0x11223344) ret addr: 0x942
[INFO]: xxx_fclose(stream=0x11223344) ret addr: 0x958
[INFO]: xxx___libc_start_main(main=0xa43, argc=0x2, ubp_av=0x13ff9d, init=0xac0, fini=0xb30, rtld_fini=0x0, stack_end=0x13ff88) ret addr: 0x73a
[INFO]: xxx_fopen(path=0x13ffc5, mode=0xd60) ret addr: 0x90a
[INFO]: xxx_fread(ptr=0x13fe30, size=0x1, nmemb=0x100, stream=0x11223345) ret addr: 0x942
[INFO]: xxx_fclose(stream=0x11223345) ret addr: 0x958
[INFO]: xxx___libc_start_main(main=0xa43, argc=0x2, ubp_av=0x13ff9d, init=0xac0, fini=0xb30, rtld_fini=0x0, stack_end=0x13ff88) ret addr: 0x73a
[INFO]: xxx_fopen(path=0x13ffc5, mode=0xd60) ret addr: 0x90a
[INFO]: xxx_fread(ptr=0x13fe30, size=0x1, nmemb=0x100, stream=0x11223346) ret addr: 0x942
[INFO]: xxx_fclose(stream=0x11223346) ret addr: 0x958
[INFO]: xxx___libc_start_main(main=0xa43, argc=0x2, ubp_av=0x13ff9d, init=0xac0, fini=0xb30, rtld_fini=0x0, stack_end=0x13ff88) ret addr: 0x73a
[INFO]: xxx_fopen(path=0x13ffc5, mode=0xd60) ret addr: 0x90a
[INFO]: xxx_fread(ptr=0x13fe30, size=0x1, nmemb=0x100, stream=0x11223347) ret addr: 0x942
[INFO]: xxx_fclose(stream=0x11223347) ret addr: 0x958
[INFO]: xxx___libc_start_main(main=0xa43, argc=0x2, ubp_av=0x13ff9d, init=0xac0, fini=0xb30, rtld_fini=0x0, stack_end=0x13ff88) ret addr: 0x73a
[INFO]: xxx_fopen(path=0x13ffc5, mode=0xd60) ret addr: 0x90a
[INFO]: xxx_fread(ptr=0x13fe30, size=0x1, nmemb=0x100, stream=0x11223348) ret addr: 0x942
[INFO]: xxx_fclose(stream=0x11223348) ret addr: 0x958
[INFO]: xxx___libc_start_main(main=0xa43, argc=0x2, ubp_av=0x13ff9d, init=0xac0, fini=0xb30, rtld_fini=0x0, stack_end=0x13ff88) ret addr: 0x73a
[INFO]: xxx_fopen(path=0x13ffc5, mode=0xd60) ret addr: 0x90a
[INFO]: xxx_fread(ptr=0x13fe30, size=0x1, nmemb=0x100, stream=0x11223349) ret addr: 0x942
[INFO]: xxx_fclose(stream=0x11223349) ret addr: 0x958
[INFO]: xxx___libc_start_main(main=0xa43, argc=0x2, ubp_av=0x13ff9d, init=0xac0, fini=0xb30, rtld_fini=0x0, stack_end=0x13ff88) ret addr: 0x73a
[INFO]: xxx_fopen(path=0x13ffc5, mode=0xd60) ret addr: 0x90a
[INFO]: xxx_fread(ptr=0x13fe30, size=0x1, nmemb=0x100, stream=0x1122334a) ret addr: 0x942
[INFO]: xxx_fclose(stream=0x1122334a) ret addr: 0x958
[INFO]: xxx___libc_start_main(main=0xa43, argc=0x2, ubp_av=0x13ff9d, init=0xac0, fini=0xb30, rtld_fini=0x0, stack_end=0x13ff88) ret addr: 0x73a
[INFO]: xxx_fopen(path=0x13ffc5, mode=0xd60) ret addr: 0x90a
[INFO]: xxx_fread(ptr=0x13fe30, size=0x1, nmemb=0x100, stream=0x1122334b) ret addr: 0x942
[INFO]: xxx_fclose(stream=0x1122334b) ret addr: 0x958
[INFO]: xxx___libc_start_main(main=0xa43, argc=0x2, ubp_av=0x13ff9d, init=0xac0, fini=0xb30, rtld_fini=0x0, stack_end=0x13ff88) ret addr: 0x73a
[INFO]: xxx_fopen(path=0x13ffc5, mode=0xd60) ret addr: 0x90a
[INFO]: xxx_fread(ptr=0x13fe30, size=0x1, nmemb=0x100, stream=0x1122334c) ret addr: 0x942
[INFO]: xxx_fclose(stream=0x1122334c) ret addr: 0x958
[INFO]: xxx___libc_start_main(main=0xa43, argc=0x2, ubp_av=0x13ff9d, init=0xac0, fini=0xb30, rtld_fini=0x0, stack_end=0x13ff88) ret addr: 0x73a
[INFO]: xxx_fopen(path=0x13ffc5, mode=0xd60) ret addr: 0x90a
[INFO]: xxx_fread(ptr=0x13fe30, size=0x1, nmemb=0x100, stream=0x1122334d) ret addr: 0x942
[INFO]: xxx_fclose(stream=0x1122334d) ret addr: 0x958
[INFO]: xxx___libc_start_main(main=0xa43, argc=0x2, ubp_av=0x13ff9d, init=0xac0, fini=0xb30, rtld_fini=0x0, stack_end=0x13ff88) ret addr: 0x73a
[INFO]: xxx_fopen(path=0x13ffc5, mode=0xd60) ret addr: 0x90a
[INFO]: xxx_fread(ptr=0x13fe30, size=0x1, nmemb=0x100, stream=0x1122334e) ret addr: 0x942
[INFO]: xxx_fclose(stream=0x1122334e) ret addr: 0x958
[INFO]: xxx___libc_start_main(main=0xa43, argc=0x2, ubp_av=0x13ff9d, init=0xac0, fini=0xb30, rtld_fini=0x0, stack_end=0x13ff88) ret addr: 0x73a
[INFO]: xxx_fopen(path=0x13ffc5, mode=0xd60) ret addr: 0x90a
[INFO]: xxx_fread(ptr=0x13fe30, size=0x1, nmemb=0x100, stream=0x1122334f) ret addr: 0x942
[INFO]: xxx_fclose(stream=0x1122334f) ret addr: 0x958
[INFO]: xxx___libc_start_main(main=0xa43, argc=0x2, ubp_av=0x13ff9d, init=0xac0, fini=0xb30, rtld_fini=0x0, stack_end=0x13ff88) ret addr: 0x73a
[INFO]: xxx_fopen(path=0x13ffc5, mode=0xd60) ret addr: 0x90a
[INFO]: xxx_fread(ptr=0x13fe30, size=0x1, nmemb=0x100, stream=0x11223350) ret addr: 0x942
[INFO]: xxx_fclose(stream=0x11223350) ret addr: 0x958
Traceback (most recent call last):
  File "dse_crackme.py", line 299, in <module>
    assert found == True
AssertionError
```

I haven't been able to repro this locally yet; and I'm unsure if it's something related to concurrent access to the cache - maybe you guys have ideas?

Cheers

0vercl0k commented:
I've run the CI 12 times (2x6 matrix) and everything is green - looks solid to me :).

https://ci.appveyor.com/project/0vercl0k/miasm-nc0e9/history

Cheers

commial commented:
> I've run the CI 12 times (2x6 matrix) and everything is green - looks solid to me :).

Looks good to me too :smiley:

commial commented on example/symbol_exec/depgraph.py:
```diff
@@ -103,7 +103,7 @@
             for element in sol.constraints:
                 try:
-                    result = hex(sol.constraints[element].as_long())
+                    result = hex(sol.constraints[element].as_long()).strip('L')
```
[cosmetic] Probably a better way to do this would be to use the "0x%x" % sol.constraints[element].as_long() form (probably more explicit than the "L" removing of `hex` artifact)

0vercl0k replied:
Done!

serpilliere commented on .appveyor.yml:
```diff
@@ -0,0 +1,61 @@
+      PLATFORM_TOOLSET: v141
+      PYTHON: c:\Python27
+      Z3_INCLUDE_DIRS: c:\projects\triton\build\z3-4.6.0-x86-win\include
+      Z3_LIBRARIES: c:\projects\triton\build\z3-4.6.0-x86-win\bin\libz3.lib
```
Yup
I didn't catch this:
Which part is using the environment variable `Z3_include_dirs` & `z3_libraries` ?

0vercl0k replied:
Shame on me! I've fixed this.

commial commented on test/test_all.py:
```diff
@@ -99,6 +102,8 @@ def __init__(self, script, jitter ,*args, **kwargs):
         if jitter in blacklist.get(script, []):
             continue
         tags = [TAGS[jitter]] if jitter in TAGS else []
+        if is_win and script.endswith("mn_div.py"):
```
I see the `mn_div` test is bypassed on Windows, is there a specific reason?

0vercl0k replied:
Yes!

I believe I'd need libLLVM for this test to work and unfortunately libLLVM isn't a binary that is shipped in the clang/llvm Windows package they ship.

So I'd need to probably compile it myself in the CI which I haven't tried out yet. We could open an issue for it and assigning it to me if you want?

serpilliere commented on miasm2/jitter/JitCore.c:
```diff
@@ -156,26 +155,6 @@ void MEM_LOOKUP_INT_BN_TO_PTR(JitCpu* jitcpu, int size, uint64_t addr, char* ptr
 	memcpy(ptr, (char*)&ret, size / 8);
 }
 
-void __attribute__((weak)) MEM_WRITE_08(JitCpu* jitcpu, uint64_t addr, uint8_t src)
```
Hum. The use of the weak attribute was to allow default symbol (with default dehavior) for the mem_lookup_xx and mem_write_xx.
In some 'special' architectures, the symbol may be redefined: For example, in arm, we do not want to flush the jitted blocks (ie instruction cache) directly after a memory write.

The problem here is my fault: when I modified the API to use bignum, I totally forgotten to use the weak attribute for the new api, which doesn't use those functions anymore (mem_write_XX) but uses mem_write_bn_xxx instead.

So here, on x86 the behavior is ok, but it may fail on arm architecture.
I whish I had written a regression test for the ARM automodifying code...

I will look at this closely at the end of the week!

0vercl0k replied:
Let me know what you want me to do here :). There's not a real equivalent AFAIK in MSVC but there's https://msdn.microsoft.com/en-us/library/5tkz6s71.aspx that could help, maybe :).

serpilliere replied:
I will have a look at this in the next days: I have to do the pro&cons between weak symbols and codegen support... Next weekend may bring us the answer :smiley:

0vercl0k commented:
@commial I initially assumed that setuptools would be needed, but I've reverted it and it works fine :) thanks!

serpilliere commented:
Erf, the last pr added some conflicts :smiley:
I am having a last review on the code.
Can you resolve the conflict, or do you want us to to this?

0vercl0k commented:
Done :)

Cheers

Le sam. 8 sept. 2018 à 04:38, serpilliere <notifications@github.com> a
écrit :

> Erf, the last pr added some conflicts 😃
> I am having a last review on the code.
> Can you resolve the conflict, or do you want us to to this?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cea-sec/miasm/pull/835#issuecomment-419635598>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABaHRacxZOEXYZBN0bCCyzmcl-cXBl0uks5uY6xPgaJpZM4WRTL0>
> .
>

serpilliere commented:
Here is the conclusion:
The weak is not really necessary here as each architecture needs a custom behavior on the automodified code (even if some share code) so the code can be moved from the `jitcore.c` to arch.

Anyway, after re-reading it, I have completely mess up when I did the PR for Bignum support: I totally ignore the original mecanism, so the modification has to be done anyway.

So I merge your great work, and I will do a PR to sanitize the whole memory handling mecanism later.


Thanks a *lot* for this great contribution @0vercl0k  !

aguinet commented:
About the ``__umodti3`` symbol, it is provided by the compiler-rt linrary. Windows used to be not supported, but with recent improvements with the overall Windows support in LLVM, chances are it is now!

If it is, I'd say you would.need to embed and compile it with Miasm native part for the jit.

0vercl0k commented:
Cheers @serpilliere :).

@aguinet awesome! I did check libLLVM and didn't see it there; but anyway it doesn't ship with the Clang package on Windows so I figured I'd have to embed it in AppVeyor/compile it myself.

I might toy with that this month if I find some time!

Cheers

aguinet commented:
libLLVM (it's actually split in multiple libraries) is used by the compiler itself. compiler-rt is a library statically linked at compile time within the final library (if necessary). It's the llvm equivalent of libgcc.

Looking at the homepage of compiler-rt, it does not seem to support windows yet. It might be possible to just copy/paste this one...

aguinet commented:
So I gave it a small try. compiler-rt's builtins compile under Windows x86-64. I've tried with LLVM 5.0 (because I already had a compiled version). llvmlite is based on LLVM6, and AFAIK will soon be based on LLVM7 when it will be out.

I'd say that a proper fix would be to have llvmlite include compiler-rt when compiling LLVM. I'll try to see if I have time this week to make a PR for this!

0vercl0k commented:
Awesome! I've looked a bit at llvmlite and their patches. Are you going to
try to get this in llvmlite or you would generate a compiler-rt library
with the proper exports and load it from the jitter?

Cheers

Le lun. 10 sept. 2018 à 00:40, Adrien Guinet <notifications@github.com> a
écrit :

> So I gave it a small try. compiler-rt's builtins compile under Windows
> x86-64. I've tried with LLVM 5.0 (because I already had a compiled
> version). llvmlite is based on LLVM6, and AFAIK will soon be based on LLVM7
> when it will be out.
>
> I'd say that a proper fix would be to have llvmlite include compiler-rt
> when compiling LLVM. I'll try to see if I have time this week to make a PR
> for this!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cea-sec/miasm/pull/835#issuecomment-419817956>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABaHRXJcvuQiRHbLyMFTFCi9c9sv9F_Nks5uZhdqgaJpZM4WRTL0>
> .
>

0vercl0k commented:
That makes sense! (sorry for the lag)

Le mar. 11 sept. 2018 à 01:25, Adrien Guinet <notifications@github.com> a
écrit :

> I was thinking that if wa can manage to get this working directly within
> llvmlite, it would be a nice fix for everyone using llvmlite :)
>
> If this can't happen (for whatever reasons), I'd say that generating a
> compiler-rt library statically linked into Miasm native library can be a
> solution.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cea-sec/miasm/pull/835#issuecomment-420190586>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABaHRY2Umm_PGYDiAf_ZBic9XK-HBMvXks5uZ3N9gaJpZM4WRTL0>
> .
>
